pkg/runner/run_context.go

```diff
@@ -275,19 +275,30 @@ func (rc *RunContext) steps() []*model.Step {
 
 // Executor returns a pipeline executor for all the steps in the job
 func (rc *RunContext) Executor() common.Executor {
-	return newJobExecutor(rc).If(rc.isEnabled)
+	return newJobExecutor(rc, &stepFactoryImpl{}, rc).If(rc.isEnabled)
 }
 
 // Executor returns a pipeline executor for all the steps in the job
 func (rc *RunContext) CompositeExecutor() common.Executor {
 	steps := make([]common.Executor, 0)
 
+	sf := &stepFactoryImpl{}
+
 	for i, step := range rc.Composite.Runs.Steps {
 		if step.ID == "" {
 			step.ID = fmt.Sprintf("%d", i)
 		}
+
+		// create a copy of the step, since this composite action could
+		// run multiple times and we might modify the instance
 		stepcopy := step
-		stepExec := rc.newStepExecutor(&stepcopy)
+
+		step, err := sf.newStep(&stepcopy, rc)
+		if err != nil {
+			return common.NewErrorExecutor(err)
+		}
+		stepExec := common.NewPipelineExecutor(step.pre(), step.main(), step.post())
+
```

Comment:
Why exactly is that wrong? In this PR all `pre()` and `post()` functions are empty stubs.
So on code besides a call is executed.

Reply:
You mean that the current approach does not work for composite actions?

Reply:
> You mean that the current approach does not work for composite actions?

Yes it doesn't work for composite actions, better skip them like before instead of calling them in a random order.

> pre() and post() functions are empty stubs.

Yes in this PR it is **still** ok, but in the following PR ( post action support ) it becomes a problem.

> Why exactly is that wrong?

You immediately call the post action handler, after each step within composite and don't run them in the composite post handler.

The main handler of composite never calls the pre executor, only the pre handler of composite does this if it is an remote action.

Reply:
Good points. I'll update the PR to integrate that with composites.

Reply:
> Yes it doesn't work for composite actions, better skip them like before instead of calling them in a random order.

This is a bigger preparation for pre/post steps to execute. I don't see random order here, but we are calling all pre-steps before all main steps before all post steps in order of definition in the yaml file.
The order of post steps is corrected and inverted to the definition order (as is on GHA).

Since pre and post steps in this PR are only empty functions no harm is done.
The way to implement this in the next PR (which is draft and not yet ready) should be to call all pre-steps of a composite action in it's pre step. And all post steps of a composite action it it's post step.
Like this:

```yaml
jobs:
  job:
    steps:
      - id: step1
        uses: some/action@v1
      - id: step2
        uses: some-other/action@v1
```

If `some/action` and `some-other/action` are composite with both pre and post steps then we will get this execution order:

- step1 pre
  - call all pre steps which are defined by actions inside `some/action`
- step2 pre
  - call all pre steps which are defined by actions inside `some-other/action`
- step1 main
- step2 main
- step2 post
  - call all post steps (in reverse order) which are defined by actions inside `some-other/action`
- step1 post
  - call all post steps (in reverse order) which are defined by actions inside `some/action`

Reply:
@ChristopherHX Do you see any error in what I described in my previous comment?

Reply:
> I don't see random order here

This is all about composite, in job level it works.

> Since pre and post steps in this PR are only empty functions no harm is done.

I agree for this PR, but I still prefer to change that specfic line related to composite. Till you implemented it for composite actions correctly.
```suggestion
		stepExec := step.main()

```

No changes needed in this PR in my point of view.